modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTable.java

```diff
@@ -0,0 +1,186 @@
+            addEventsLimitCopy.add(streamEvent);
+            sizeAfterAdding++;
+            streamEvent = streamEvent.getNext();
+            if (sizeAfterAdding == maxSize || streamEvent.getNext() == null) {
+                break;
+            }
+        }
+        this.add(addEventsLimitCopy);
+    }
+
+    public void addUptoMaxSize(ComplexEventChunk<StreamEvent> addingEventChunk) {
+        ComplexEventChunk<StreamEvent> addingEventChunkForCache = new ComplexEventChunk<>(true);
+        addingEventChunk.reset();
+        while (addingEventChunk.hasNext()) {
+            StreamEvent event = addingEventChunk.next();
+            addRequiredFieldsToDataForCache(addingEventChunkForCache, event, siddhiAppContext, cacheExpiryEnabled);
+        }
+        super.add(addingEventChunkForCache);
+        while (this.size() > maxSize) {
+            this.deleteOneEntryUsingCachePolicy();
+        }
+    }
+
+    public void updateOrAddWithMaxSize(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
+                                       CompiledCondition compiledCondition,
+                                       CompiledUpdateSet compiledUpdateSet,
+                                       AddingStreamEventExtractor addingStreamEventExtractor, int maxTableSize) {
+        ComplexEventChunk<StateEvent> updateOrAddingEventChunkForCache = new ComplexEventChunk<>(true);
+        updateOrAddingEventChunk.reset();
+        while (updateOrAddingEventChunk.hasNext()) {
+            StateEvent event = updateOrAddingEventChunk.next();
+            addRequiredFieldsToDataForCache(updateOrAddingEventChunkForCache, event, siddhiAppContext,
+                    cacheExpiryEnabled);
+        }
+
+        readWriteLock.writeLock().lock();
+        TableState state = stateHolder.getState();
+        try {
+            ComplexEventChunk<StreamEvent> failedEvents = ((Operator) compiledCondition).tryUpdate(
+                    updateOrAddingEventChunkForCache,
+                    state.getEventHolder(),
+                    (InMemoryCompiledUpdateSet) compiledUpdateSet,
+                    addingStreamEventExtractor);
+            if (failedEvents != null && failedEvents.getFirst() != null) {
+                state.getEventHolder().add(failedEvents);
+            }
+            while (this.size() > maxSize) {
+                this.deleteOneEntryUsingCachePolicy();
+            }
+        } finally {
+            stateHolder.returnState(state);
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    public void deleteAll() {
+        stateHolder.getState().getEventHolder().deleteAll();
+    }
+
+    abstract void addRequiredFieldsToCacheTableDefinition(TableDefinition cacheTableDefinition,
+                                                          boolean cacheExpiryEnabled);
+
+    public abstract void deleteOneEntryUsingCachePolicy();
+
+    public void addRequiredFieldsToDataForCache(Object addingEventChunkForCache, Object event,
```

Comment:
Change this method to generate the eventWithRequriedFields, and do not pass the addingEventChunkForCache into this.